### PR TITLE
panel_6

### LINKED ISSUE #154: panel_6
Description: Paneling four plots and demonstrating adding white space between panels

Script URL: https://www.ncl.ucar.edu/Applications/Scripts/panel_6.ncl
Plot URL: https://www.ncl.ucar.edu/Applications/Images/panel_6_1_lg.png
https://www.ncl.ucar.edu/Applications/Images/panel_6_2_lg.png

Category: Panel

### PR DESCRIPTION
**Pull request by hCraker**
Closes #154 
image

### CODE REVIEW

**Comment by hCraker** [comment]
I'm having some issues with the color maps. The data for the bottom plots (representing salinity) has values in the range of 30 to 40 psu while the NCL plot is showing values that are much closer to 0 and even negative. I don't see any obvious data manipulation in the NCL script. Do either of you have an idea of what might be causing this? @clairefio @michaelavs

**Comment by michaelavs** [comment]
> I'm having some issues with the color maps. The data for the bottom plots (representing salinity) has values in the range of 30 to 40 psu while the NCL plot is showing values that are much closer to 0 and even negative. I don't see any obvious data manipulation in the NCL script. Do either of you have an idea of what might be causing this? @clairefio @michaelavs
> 
> Additionally, I'm using the same color map as the NCL version but the colors don't match even on the top two plots which have the same data as the upper plots on the NCL example.

My first thought is there may be a units problem, but I don't think that would be the solution based on the current values being plotted and that you were able to get at least 2/4 plotted. Another option may be how refined the data is with the number of levels used. 
For the actual coloring problem, I wonder if it's tied to the data range in the sense that the yellow color that should be the max from NCL is what Python may recognize only as a middle color and therefore sets the range as blue to red not blue to yellow as expected. If need be, I think it's more than fine to use a different color scheme for this if the issue doesn't resolve itself once the first issue is resolved for this

**Comment by hCraker** [comment]
> My first thought is there may be a units problem, but I don't think that would be the solution based on the current values being plotted and that you were able to get at least 2/4 plotted. Another option may be how refined the data is with the number of levels used.

Can you elaborate on what you mean by "how refined" the data is? I've tried plotting all of the levels specified by `z_t` in the NetCDF file, and none of them look remotely similar to what the NCL script is outputting.

**Comment by michaelavs** [comment]
> > My first thought is there may be a units problem, but I don't think that would be the solution based on the current values being plotted and that you were able to get at least 2/4 plotted. Another option may be how refined the data is with the number of levels used.
> 
> Can you elaborate on what you mean by "how refined" the data is? I've tried plotting all of the levels specified by `z_t` in the NetCDF file, and none of them look remotely similar to what the NCL script is outputting.

I'm not if it will be the same with this projection style, but I know when I used WRF Python for a class, if you didn't set enough levels for the contours, it would make the projection either one singular color or would not reflect the correct amount of contours on the map. So it's really more of just how refined the projection is: more levels being more "refined" and able to capture the changes better between the contours

**Comment by hCraker** [comment]
@michaelavs I know you have worked with Polar Stereographic projections before. Have you figured out how to draw the longitude labels around the map edge? This is what I currently have after modifying the code you used for #68 

image

**Comment by michaelavs** [comment]
> @michaelavs I know you have worked with Polar Stereographic projections before. Have you figured out how to draw the longitude labels around the map edge? This is what I currently have after modifying the code you used for #68
> 
> image

We had to use an incredibly hard coded work around which was:
`# Manipulate longitude labels (0, 30 E, 60 E, ..., 30 W, etc.)
ticks = np.arange(0, 210, 30)
etick = ['0'] + [r'%d$^\circ$E' %
                 tick for tick in ticks if (tick != 0) & (tick != 180)] + ['180']
wtick = [
    r'%d$^\circ$W' %
    tick for tick in ticks if (
        tick != 0) & (
            tick != 180)]
labels = etick + wtick
xticks = [-0.8, 28, 58, 89.1, 120, 151, 182.9, -36, -63, -89, -114, -140, -151]
yticks = [-3] + [-2] + [-1] + [-1] * 2 + [-1] + [-3] + [-7] + [-7] * 3 + [-7]


for xtick, ytick, label in zip(xticks, yticks, labels):
    ax.text(xtick, ytick, label, transform=ccrs.Geodetic())
`
The way to move the labels around (which I just had to experiment with, there wasn't necessarily a rhyme or reason to it) was by modifying the 'yticks' parameters. Something similar may work with the 'xticks' parameters if that is necessary but I don't think I had to adjust those at all

**Comment by hCraker** [comment]
Thanks @michaelavs. That was the code I was trying to modify. If hard coded values are necessary, I will give them a try.

**Comment by hCraker** [comment]
@michaelavs I have found an easy way to draw the longitude labels without hard coding anything. I tried it out on your polar projection plot #68 and it worked there too, so I think this is a robust solution. If you don't mind, I'd like to open up a issue and PR to change the polar_1 code to use this method.

**Comment by hCraker** [comment]
image

**Comment by michaelavs** [comment]
> @michaelavs I have found an easy way to draw the longitude labels without hard coding anything. I tried it out on your polar projection plot #68 and it worked there too, so I think this is a robust solution. If you don't mind, I'd like to open up a issue and PR to change the polar_1 code to use this method.

Definitely! I think we had originally planned on returning to that projection to update it if/when a solution came about for the hard coded side, so re-opening an issue on it is more than ok to me!

**Comment by hCraker** [comment]
Plot without any added white space:
image

Plot with added white space:
image

File: Plots/Panels/NCL_panel_6.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+import cartopy.feature as cfeature
+import matplotlib.pyplot as plt
+import matplotlib.ticker as mticker
+import matplotlib.contour as mcontour
+from mpl_toolkits.axes_grid1.inset_locator import inset_axes
+import numpy as np
+import xarray as xr
+
+import geocat.datafiles as gdf
+from geocat.viz import cmaps as gvcmaps
```
Comment (michaelavs) [inline Plots/Panels/NCL_panel_6.py:32]:
```suggestion

```
This line is unused in the code

File: Plots/Panels/NCL_panel_6.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+            this if the western edge of the plot is cut off. Defaults to 0.
+        res (:class:`int`):
+            The size of the incrementation for vertices in degrees. Default is
+            a vertex every one degree of longitude. A higher number results in
+            a lower resolution boundary.
+    """
+    import cartopy.crs as ccrs
+    import matplotlib.path as mpath
+    import matplotlib.patches as mpatches
+    import numpy as np
```
Comment (michaelavs) [inline Plots/Panels/NCL_panel_6.py:92]:
```suggestion

```

File: Plots/Panels/NCL_panel_6.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+            A constant to be subtracted from the first entry in lon_range. Use
+            this if the western edge of the plot is cut off. Defaults to 0.
+        res (:class:`int`):
+            The size of the incrementation for vertices in degrees. Default is
+            a vertex every one degree of longitude. A higher number results in
+            a lower resolution boundary.
+    """
+    import cartopy.crs as ccrs
+    import matplotlib.path as mpath
+    import matplotlib.patches as mpatches
```
Comment (michaelavs) [inline Plots/Panels/NCL_panel_6.py:91]:
```suggestion

```

**Comment by hCraker** [comment]
> Code looks nice! Plots are looking good except one point: In the first plot, the colorbars of the left column overlaps with the plot of the right column as in the red ellipse below:
> 
> image

That is the intent with this example. The first image shows what the plot looks like without added white space just like in the NCL example. The difference is that `matplotlib` has a default arrangement that leads to overlap while NCL's default looks nicer.

**Comment by erogluorhan** [comment]
> That is the intent with this example. The first image shows what the plot looks like without added white space just like in the NCL example. The difference is that `matplotlib` has a default arrangement that leads to overlap while NCL's default looks nicer. Should I change this plot?

Actually, I thought about that would be intentional, but yes, I am confused since NCL plot has no overlap. I don ot know if we should have this as is with the overlap, or get rid of the overlap but not put a large whitespace as in the 2nd plots? Any thoughts @clyne @jukent ?

**Comment by hCraker** [comment]
I mis-clicked and didn't intend to close this.

**Comment by jukent** [comment]
> > That is the intent with this example. The first image shows what the plot looks like without added white space just like in the NCL example. The difference is that `matplotlib` has a default arrangement that leads to overlap while NCL's default looks nicer. Should I change this plot?
> 
> Actually, I thought about that would be intentional, but yes, I am confused since NCL plot has no overlap. I don ot know if we should have this as is with the overlap, or get rid of the overlap but not put a large whitespace as in the 2nd plots? Any thoughts @clyne @jukent ?

This is an interesting issue because as @hCraker points out the two examples are one without spacing and one fixing it. I think we have 3 options. 1 - leave it as is and use the overlap as motivation in the documentation for why we'd want to space out the plots. 2 - use minimum spacing to avoid overlap in the first plot (leaving both examples kind of redundant to each other. 3 - only showing the second plot that does not have overlap.

**Comment by hCraker** [comment]
> This is an interesting issue because as @hCraker points out the two examples are one without spacing and one fixing it. I think we have 3 options. 1 - leave it as is and use the overlap as motivation in the documentation for why we'd want to space out the plots. 2 - use minimum spacing to avoid overlap in the first plot (leaving both examples kind of redundant to each other). 3 - only show the second plot that does not have overlap.

Option 1 is what I was trying to do. If we only leave the second plot with the fixed spacing, that circumvents the opportunity to teach users how to control panel spacing. I think showing users this feature is the intent with the example given that the NCL script comments state that "Adding white space around paneled plots" is an illustrated concept. Thoughts? @erogluorhan @jukent @clyne

**Comment by erogluorhan** [comment]
> Option 1 is what I was trying to do. If we only leave the second plot with the fixed spacing, that circumvents the opportunity to teach users how to control panel spacing. I think showing users this feature is the intent with the example given that the NCL script comments state that "Adding white space around paneled plots" is an illustrated concept. Thoughts? @erogluorhan @jukent @clyne

Taking into attention the original intent, I think it is better to kepp whatever @hCraker is already implemented.

**Comment by clyne** [comment]
> > Option 1 is what I was trying to do. If we only leave the second plot with the fixed spacing, that circumvents the opportunity to teach users how to control panel spacing. I think showing users this feature is the intent with the example given that the NCL script comments state that "Adding white space around paneled plots" is an illustrated concept. Thoughts? @erogluorhan @jukent @clyne
> 
> Taking into attention the original intent, I think it is better to keep whatever @hCraker is already implemented.

Works for me.

**Pull request merged**